src/QirRuntime/public/CoreTypes.hpp

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+// The core types will be exposed in the C-interfaces for interop, thus no
+// namespaces or scoped enums can be used to define them.
+
+/*==============================================================================
+  Helper types
+==============================================================================*/
+enum TernaryBool

```

Comment:
Just curious: when is this used? The Boolean type in QIR, at least, is always either true or false, never pending or undefined.

Reply:
Yes, QIR doesn't support deferred results but from the conversations with the researches the ability to defer measurements came up as a desirable feature. To be able to experiment with it but to avoid throwing from the C-API the enum was introduced. I'm not too happy about it, but thinking through the design here was one of the tasks that itself got, no pun intended, deferred.

Reply:
That sounds like a solution rather than a need that was communicated to you by researchers. Could you please clarify what exactly the scenarios are that need to be supported where this would be useful, and then we can make an informed call on how to best support them.

Reply:
https://ms-quantum.visualstudio.com/Quantum%20Program/_workitems/edit/14449 used IonQ's result placeholders as an example when it might not be possible to get value from a measurement and input from Vadym re avoiding throwing from the C-API. I don't think we've ever had a conversation about the scenarios the native runtime is supposed to support. There is no spec or design document for it, so in a way, all code in this project is a solution. To something.